### PR TITLE
AIP-61 - Support task adoption for hybrid executors

### PR DESCRIPTION
**Pull request by o-nikolas**
Sort the set of tasks that are up for adoption by the executor they're configured to run on (if any) and send them to the appropriate executor for adoption.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/executors/executor_loader.py [hunk 1]
```diff
@@ -202,10 +202,10 @@ def lookup_executor_name_by_str(cls, executor_name_str: str) -> ExecutorName:
         elif executor_name := _module_to_executors.get(executor_name_str):
             return executor_name
         else:
-            raise AirflowException(f"Unknown executor being loaded: {executor_name}")
+            raise AirflowException(f"Unknown executor being loaded: {executor_name_str}")
```
Comment (o-nikolas) [inline airflow/executors/executor_loader.py:205]: This was an unrelated bug fix, just a one liner

File: airflow/executors/executor_loader.py [hunk 2]
```diff
@@ -202,10 +202,10 @@ def lookup_executor_name_by_str(cls, executor_name_str: str) -> ExecutorName:
...
 
     @classmethod
-    def load_executor(cls, executor_name: ExecutorName | str) -> BaseExecutor:
+    def load_executor(cls, executor_name: ExecutorName | str | None) -> BaseExecutor:
```
Comment (o-nikolas) [inline airflow/executors/executor_loader.py:208]: We now except None (which means In some cases that there was no specified executor, so in that case we should return the default executor).

File: airflow/jobs/scheduler_job_runner.py [hunk 3]
```diff
@@ -1831,3 +1836,16 @@ def _orphan_unreferenced_datasets(self, session: Session = NEW_SESSION) -> None:
...
+
+    def _executor_to_tis(self, tis: list[TaskInstance]) -> dict[BaseExecutor, list[TaskInstance]]:
+        """Organize TIs into lists per their respective executor."""
+        _executor_to_tis: defaultdict[BaseExecutor, list[TaskInstance]] = defaultdict(list)
+        executor: str | None
+        for ti in tis:
+            if ti.executor:
+                executor = str(ti.executor)
+            else:
+                executor = None
```
Comment (o-nikolas) [inline airflow/jobs/scheduler_job_runner.py:1848]: This bit of weirdness is just to make mypy types happy. The ti.executor field type comes up as a SQLA Column type, so if it's not None I cast to a string and if it's None I just make it None again (because mypy is stupid and harms my life).

File: tests/executors/test_executor_loader.py [hunk 4]
```diff
@@ -301,3 +302,21 @@ def test_validate_database_executor_compatibility_sqlite(self, monkeypatch, exec
         monkeypatch.delenv("_AIRFLOW__SKIP_DATABASE_EXECUTOR_COMPATIBILITY_CHECK")
         with expectation:
             ExecutorLoader.validate_database_executor_compatibility(executor)
+
+    def test_load_executor(self):
```
Comment (o-nikolas) [inline tests/executors/test_executor_loader.py:306]: I touched this method above and noticed it had no tests (at least I can't find them) so I added a couple new ones here.

Comment (ferruzzi) [reply tests/executors/test_executor_loader.py:306]: Nice addition

**Review by ferruzzi — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Review by Taragolis — APPROVED** [review]
